### PR TITLE
Implement resource based permissions control.

### PR DESCRIPTION
**Pull request by ycheng-aa**
1. Provide 3 APIs, resource-permissions,
   current-user-resource-permissions and
   group-resource-permissions.
2. Resources are APIs which need to be included in permission
   control. Permissions are creat, read, update, delete. PDC
   will collect resource permissions accroding to router
   automatically.
3. resource-permissions API provide list function to
   list all available resource permissions.
   current-user-resource-permissions list current
   user's resource permissions.
4. Through group-resource-permissions, resource permission
   could be assigned to or removed from a group. Users in
   that group will be affected.
5. Prevent non-superuser from accessing admin page.

JIRA: PDC-1384

### CODE REVIEW

File: pdc/apps/auth/migrations/0004_auto_20160512_1051.py [hunk 1]
```diff
@@ -0,0 +1,56 @@
+# -*- coding: utf-8 -*-
+from __future__ import unicode_literals
+
+from django.db import models, migrations
+
+
+def create_action(apps, schema_editor):
+    """Migrate contacts stored directly on release components."""
```
Comment (lubomir) [inline pdc/apps/auth/migrations/0004_auto_20160512_1051.py]: This comment does not seem to match the method.

File: pdc/apps/auth/views.py [hunk 2]
```diff
@@ -527,3 +530,134 @@ def list(self, request):
...
+        resource_permission_set = set([])
+        for group_resource_permission in queryset:
+            resource_permission_set.add(group_resource_permission.resource_permission)
+        serializer = serializers.ResourcePermissionSerializer(resource_permission_set, many=True)
+        return Response(serializer.data)
+
+
+class GroupResourcePermissionViewSet(common_viewsets.PDCModelViewSet):
+    """
+    This end-point provides t group resource permissions.
```
Comment (lubomir) [inline pdc/apps/auth/views.py]: Stray *t*.

File: pdc/apps/auth/views.py [hunk 3]
```diff
@@ -527,3 +530,134 @@ def list(self, request):
...
+        __URL__: $LINK:groupresourcepermissions-list$
+
+        __Query params__:
+
+        %(FILTERS)s
+
+
+        __Response__:
+
+
```
Comment (lubomir) [inline pdc/apps/auth/views.py]: Is response description missing here?

File: pdc/apps/utils/middleware.py [hunk 4]
```diff
@@ -27,3 +34,50 @@ def process_response(self, request, response):
...
+
+
+class RestrictAdminMiddleware(object):
+    """
+    Restricts access to the admin page to only logged-in users with a certain user-level.
+    """
+    def process_request(self, request):
+        if request.user.is_authenticated() and request.path.startswith(reverse('admin:index')):
+            if not (request.user.is_active and request.user.is_superuser):
+                return HttpResponseRedirect("/%s%s/" % (settings.REST_API_URL, settings.REST_API_VERSION))
```
Comment (lubomir) [inline pdc/apps/utils/middleware.py]: Why is this needed? The behaviour up to now was to allow login to admin section for users who are marked as *staff*.

File: pdc/apps/auth/views.py [hunk 5]
```diff
@@ -527,3 +530,134 @@ def list(self, request):
...
+    def list(self, request):
+        """
+        Get information about current user.
+
+        __Method__: `GET`
+
+        __URL__: $LINK:currentuserpermissions-list$
+
+        __Response__:
+            [
```
Comment (lubomir) [inline pdc/apps/auth/views.py]: This does not render correctly. Maybe a blank line is missing?

**Comment by lubomir** [comment]
The code looks reasonable to me. I'm not a big fan of adding another dependency, but if it makes things easier, then why not.

I think it would be nice to have the new urls grouped under the `auth/` prefix. How about removing the `resource` keyword from the URL too? Does it add any value?

A couple notes from my testing:
 * Running a development server with this patch suddenly requires login to access any API. How can the server be configured to not require any authentication for read access? This is a requirement for deploying in Fedora infrastructure.
 * Even though I'm an admin, visiting `/current-user-resource-permission` does not show anything. Furthermore, why add another end-point here instead of merging this into existing `/auth/current-user`?
 * Urls in documentation for new APIs don't show correctly.
 * Group resource permissions endpoint does not have *Create* tab, but it has *Bulk Create*.

File: pdc/apps/utils/middleware.py (hunk 4 above)
Comment (ycheng-aa) [inline pdc/apps/utils/middleware.py]: I thought if we grant all model permissions to all groups, then API permissions could work all by itself. PDC will not have the situation that it grant a group a API's permission, but still can't work because lack of model permission. But it will bring the risk that user could manipulate model directly to avoid permission control.

I think you mentioned we should remove model permissions. But I am not sure what it means.  Did you mean remove `django.contrib.auth.backends.ModelBackend` from `AUTHENTICATION_BACKENDS`? I tested to remove it and it looks we can't login admin page.

File: pdc/apps/auth/migrations/0004_auto_20160512_1051.py (hunk 1 above)
Comment (ycheng-aa) [inline pdc/apps/auth/migrations/0004_auto_20160512_1051.py]: Fixed in 71b3211d41ee6bd4

File: pdc/apps/auth/views.py (hunk 5 above)
Comment (ycheng-aa) [inline pdc/apps/auth/views.py]: Fixed in 71b3211d41ee6bd4

File: pdc/apps/auth/views.py (hunk 2 above)
Comment (ycheng-aa) [inline pdc/apps/auth/views.py]: Fixed in 71b3211d41ee6bd4

File: pdc/apps/auth/views.py (hunk 3 above)
Comment (ycheng-aa) [inline pdc/apps/auth/views.py]: Fixed in 71b3211d41ee6bd4

**Comment by ycheng-aa** [comment]
@lubomir  Thanks you very much for the careful review.

* I moved new endpoints under `auth/`. Merged resource permission content to `/auth/current-user`.
* For superuser, It will show all resource permissions even not explicitly assigned to him and doesn't   exist in DB.
* Fixed documentation and make sure `create` tab work.
* In settings.py, add a flag to allow anyone has read permission.

**Comment by lubomir** [comment]
Regarding the model permissions, I think it does not matter if they are used or not as long as they are not exposed to users (including admins). It seems confusing to have to different permission systems. What is the expected workflow?

**Comment by ycheng-aa** [comment]
@lubomir I am hoping we could grant all model permissions to all groups. Then use resource based permissions control only.  This is my expected workflow.
Because user could have all model permissions, grant any resource permission to a user could make sure user really has the permission. Normal user also couldn't modify model directly instead of by accessing API. 
All this just my conjecture, I am not sure if there is something wrong in the workflow.

**Comment by lubomir** [comment]
That sounds reasonable to me. If everyone has all the model permissions, are changes in the admin section still limited to admins only?

**Comment by lubomir** [comment]
Anyway, my proposal here would be to merge this pull request and deploy it on devel server. Then we can test how it behaves, talk to users and have them try it. We have quite some time until next release to fix any issues.

**Comment by ycheng-aa** [comment]
Changes you mean the page like https://pdc.host.dev.eng.pek2.redhat.com/changes/ ? If it's this page, it's not in admin section.
I am also not sure if the admin restriction is necessary. As you said, we could deploy then test and fix problems. Thanks.

**Comment by simozhan** [comment]
Looks good to me. :+1: Agree with @lubomir, merge it first and test it.
